lib/spack/spack/ci.py

```diff
@@ -638,6 +651,74 @@ def _apply_section(dest, src):
                         job["attributes"] = self.__apply_submapping(
                             job["attributes"], job["spec"], section
                         )
+            elif has_dynmapping:
+                # Unpack in indent script
+                mode = section["dynamic-mapping"].get("mode", "batch")
+                url = section["dynamic-mapping"].get("url", "http://localhost")
+                port = section["dynamic-mapping"].get("port")
+                headers = section["dynamic-mapping"].get("headers", {})
+
+                if url.startswith("env::"):
+                    url = os.environ.get(url.lstrip("env::"))
+
+                if port.startswith("env::"):
+                    port = os.environ.get(port.lstrip("env::"))
+
+                if port:
+                    url += f":{port}"
+
+                url += "/v1/kube-requests"
+                print("Dynamic Mapping URL: ", url)
+
+                def job_payload(job):
+                    job_vars = job["attributes"].get("variables")
+                    return {
+                        "hash": job_vars.get("SPACK_JOB_SPEC_DAG_HASH", ""),
+                        "arch": job_vars.get("SPACK_JOB_SPEC_ARCH", ""),
+                        "package": {
+                            "name": job_vars.get("SPACK_JOB_SPEC_PKG_NAME", ""),
+                            "version": job_vars.get("SPACK_JOB_SPEC_PKG_VERSION", ""),
+                            "variants": job_vars.get("SPACK_JOB_SPEC_VARIANTS", ""),
+                        },
+                        "compiler": {
+                            "name": job_vars.get("SPACK_JOB_SPEC_COMPILER_NAME", ""),
+                            "version": job_vars.get("SPACK_JOB_SPEC_COMPILER_VERSION", ""),
+                        },
+                    }
+
+                opener = build_opener(HTTPHandler)
+                # TODO: timeout, retry policy, etc. need to be implemented here
+                if mode == "batch":
+                    data = []
+                    for job in jobs.values():
+                        if job["spec"]:
+                            data.append(job_payload(job))
+
+                    # TODO: allow setting message size limits
+                    data = json.dumps(data).encode("utf-8")
+                    request = Request(url, data=data, headers=headers, method="POST")
+
+                    response = opener.open(request)
+                    mappings = json.loads(response.read().decode("utf-8"))
+                    for sha, job in jobs.items():
+                        if not job["spec"]:
+                            continue
+                        job["attributes"] = spack.config.merge_yaml(
+                            job.get("attributes", {}), mappings.get(sha, {})
```

Comment:
Could this allow the API to set arbitrary env variables? I wonder if we want to add an allow-list of variables that are acceptable (CPU/mem request+limit, # of build jobs)

Reply:
Also, if this is implemented, will it replace the variables set in `share/spack/gitlab/cloud_pipelines/configs/linux/ci.yaml`, like CI_JOB_SIZE?

Reply:
This is a general feature that is not specific to the core spack ci workflow, it could be that people will set up different endpoints to inject different types of variables for different schedulers/cluster types/etc.

By not having explicit restriction this feature is flexible enough to handle a number of use cases beyond dynamic resource allocation, it is simply a dynamic build job attribute mapper for the spack ci deployment to determine what they think is important to map dynamically based on what is being rebuilt, in our case it just happens to be kube requests.